frame/vamm/src/helpers/swap.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{Config, Error, Event, Pallet, SwapConfigOf, SwapOutputOf, VammMap, VammStateOf};
+use composable_maths::labs::numbers::UnsignedMath;
+use composable_traits::vamm::{AssetType, Direction, SwapOutput};
+use frame_support::{pallet_prelude::*, transactional};
+use sp_runtime::ArithmeticError;
+use std::cmp::Ordering;
+
+// TODO(Cardosaum): Document this struct:
+// - Why is it needed
+// - What's its purpose
+// - What does each field represent
+// - Why doing this way and not using tuples for example?
+struct CalculateSwapAsset<T: Config> {
+	output_amount: T::Balance,
+	input_amount: T::Balance,
+}
+
+/// Stores the result of a swap operation, containing both the new values for
+/// `base` and `quote` assets as well as how much the caller would receive (or
+/// pay) in return for this specific swap.
+pub struct ComputeSwap<T: Config> {
+	/// The new amount of `base` asset that the vamm would contain after this
+	/// swap.
+	base_asset_reserves: T::Balance,
+	/// The new amount of `quote` asset that the vamm would contain after this
+	/// swap.
+	quote_asset_reserves: T::Balance,
+	/// The total asset amount the caller will receive/pay in terms of the the
+	/// opposite asset for this swap to takes place.
+	pub swap_output: SwapOutputOf<T>,
+}
+
+impl<T: Config> Pallet<T> {
+	/// Performs runtime storage changes, effectively conducting the asset swap.
+	/// For more information about what this function does, read
+	/// [`swap`](struct.Pallet.html#method.swap) function documentation.
+	///
+	/// # Errors
+	///
+	/// * [`Error::<T>::BaseAssetReservesWouldBeCompletelyDrained`]
+	/// * [`Error::<T>::FailToRetrieveVamm`]
+	/// * [`Error::<T>::InsufficientFundsForTrade`]
+	/// * [`Error::<T>::QuoteAssetReservesWouldBeCompletelyDrained`]
+	/// * [`Error::<T>::SwappedAmountLessThanMinimumLimit`]
+	/// * [`Error::<T>::TradeExtrapolatesMaximumSupportedAmount`]
+	/// * [`ArithmeticError`](sp_runtime::ArithmeticError)
+	#[transactional]
+	pub fn do_swap(
+		config: &SwapConfigOf<T>,
+		vamm_state: &mut VammStateOf<T>,
+	) -> Result<SwapOutputOf<T>, DispatchError> {
+		// Compute new reserves of base and quote asset and swap result.
+		let ComputeSwap { base_asset_reserves, quote_asset_reserves, swap_output } =
+			Self::compute_swap(config, vamm_state)?;
+
+		// Update runtime storage.
+		VammMap::<T>::try_mutate(&config.vamm_id, |old_vamm_state| match old_vamm_state {
+			Some(v) => {
+				v.base_asset_reserves = base_asset_reserves;
+				v.quote_asset_reserves = quote_asset_reserves;
+				Ok(())
+			},
+			None => Err(Error::<T>::FailToRetrieveVamm),
+		})?;
+
+		// Deposit swap event into blockchain.
+		Self::deposit_event(Event::<T>::Swapped {
+			vamm_id: config.vamm_id,
+			input_amount: config.input_amount,
+			output_amount: swap_output,
+			input_asset_type: config.asset,
+			direction: config.direction,
+		});
+
+		// Return swap output.
```

Comment:
NIT: this comment is equally as explanatory as Ok(swap_output) so it can be deleted.

This applies to the functions below this one too